ntbs-service/DataMigration/NotificationImportService.cs

```diff
@@ -0,0 +1,126 @@
+using System.Linq;
+using System.Threading.Tasks;
+using EFAuditer;
+using ntbs_service.DataAccess;
+using ntbs_service.Models;
+using ntbs_service.Models.Enums;
+using Serilog;
+
+namespace ntbs_service.DataMigration
+{
+
+    public interface INotificationImportService
+    {
+        Task<IList<Notification>> ImportNotificationsAsync(string notificationId);
+    }
+
+    public class NotificationImportService : INotificationImportService
+    {
+        private readonly INotificationSearcher notificationSearcher;
+        private readonly INotificationRepository notificationRepository;
+        private readonly NtbsContext _context;
+
+        public NotificationImportService(INotificationSearcher notificationSearcher,
+                             INotificationRepository notificationRepository,
+                             NtbsContext context)
+        {
+            this.notificationSearcher = notificationSearcher;
+            this.notificationRepository = notificationRepository;
+            _context = context;
+        }
+
+        public async Task<IList<Notification>> ImportNotificationsAsync(string notificationId)
+        {
+            LogInformation($"Request to import Notification with Id={notificationId}");
+
+            if (notificationRepository.NotificationWithLtbrOrEtsIdExists(notificationId)) 
+            {
+                LogInformation($"Notification with Id={notificationId} already exists in database");
+                return null;
+            }
+
+            var notificationTask = notificationSearcher.Search(notificationId);
+            Task.WaitAll(notificationTask);
+            var notifications = notificationTask.Result.ToList();
+
+            LogInformation($"{notifications.Count} notifications included linked ones found to import");
+            var notificationsToSave = new List<Notification>();
+            foreach (var notification in notifications)
+            {
+                var linkedNotificationId = notification.LTBRID ?? notification.ETSID;  
+                LogInformation($"Validating notification with Id={linkedNotificationId}");
+
+                var validationErrors = GetValidationErrors(notification);
+                
+                if (validationErrors.Count() == 0)
+                {
+                    notificationsToSave.Add(notification);
+                    LogInformation($"No validation errors found");
+                }
+                else
+                {
+                    LogInformation($"{validationErrors.Count()} validation errors found:");
+                    foreach (var validationError in validationErrors)
+                    {
+                        LogInformation(validationError.ErrorMessage);
+                    }
+                }
+            }
+            LogInformation($"Importing {notificationsToSave.Count()} valid notifications");
+            var savedNotificaitons = await AddLinkedNotificationsAsync(notificationsToSave);

```

Comment:
I think we only want to save all notifications (if no validation errors) - or none.
Otherwise it could be that we import 2/3 notifications for a patient, but 1 of them will be stuck in legacy system. even if we fix the data and try to import it again, we'll have lost the group link (or worse yet, we'd import all 3 again, causing duplicates)!